### PR TITLE
fix(feature-discovery): exclude inactive tasks from top-k utility

### LINKED ISSUE #379: FixedBudgetFeatureLearner's top-k utility still averages inactive-task zero placeholders (the #275 fix landed only in the interaction learner)
## Summary

#275 / PR #276 fixed `FixedBudgetInteractionLearner`'s top-k utility so that, under `utility_task_balancing in {"active", "active_inverse_frequency"}`, inactive-task zero placeholders no longer dilute the top-k mean (mask inactive rows to `-inf`, divide by the selected active count). PR #276 touched only `interaction_features.py`; the byte-identical code in `FixedBudgetFeatureLearner` (`alberta_framework/core/feature_discovery.py`, `_output_utility_signal` and `_aggregate_task_feature_signal`) was left behind:

```python
if self._utility_aggregation == "topk":
    k = min(self._utility_top_k, self._n_tasks)
    return jnp.mean(jnp.sort(weighted_activity, axis=0)[-k:, :], axis=0)   # zeros of inactive tasks averaged in
```

The dilution factor is `(#active heads at that step) / k`, so a feature measured on a sparse-label step is systematically penalised against an identical feature measured on a dense step; `worst_active = argmin(active_scores)` then discards the wrong slot, and the same diluted utilities feed `should_promote`, the promotion pressure, and metrics 2/3/4.

## Falsifier (deterministic, pristine `main` @ `45279f3`)

```python
import jax.numpy as jnp
from alberta_framework import FixedBudgetFeatureLearner, FixedBudgetInteractionLearner
active = jnp.asarray([True, False, False, False]); ema = jnp.ones(4)
out_w = jnp.asarray([[6.0], [5.0], [5.0], [5.0]]); feats = jnp.ones(1)
for k in (1, 2, 4):
    fd = FixedBudgetFeatureLearner(n_features=1, n_tasks=4, candidate_count=0,
                                   utility_aggregation="topk", utility_top_k=k, utility_task_balancing="active")
    print("top_k", k, "feature_discovery:", fd._output_utility_signal(out_w, feats, active, ema), "expected [6.]")
```

Output on `main`:

```
top_k 1 feature_discovery: [6.]  expected [6.]
top_k 2 feature_discovery: [3.]  expected [6.]
top_k 4 feature_discovery: [1.5] expected [6.]
```

Ranking flip (3 heads, `top_k=2`, `balancing="active"`): feature A (1 active head, real signal 5.0) scores `2.5`, feature B (2 active heads, real 3.0) scores `3.0` — the discovery learner declares A the worse feature while the fixed interaction learner (5.0 vs 3.0) declares B. Through the public `update()` with only head 0 observed, utilities shrink by 0.55–0.88× when `top_k` goes 1→2 and 0.40–0.85× at `top_k=3` on the same signal.

Expected: the same semantics as #276 — under task balancing only active rows compete for the top-k and the mean divides by `min(#active, k)`; the `"none"` branch stays byte-identical.

## Dedup

#275 / PR #276 (same defect, `interaction_features.py` only — verified via `gh pr view 276 --json files`). No issue mentions `FixedBudgetFeatureLearner` or `feature_discovery` top-k; no open PR touches `feature_discovery.py`. (`core/interaction_features.py` is a registered evidence source; `core/feature_discovery.py` is not.)

Fix PR follows (failing-test-first).

### PR DESCRIPTION
**Pull request by Prabhat1308**
Closes #379.

## Issue

`FixedBudgetFeatureLearner`'s top-k utility (`_output_utility_signal`, `_aggregate_task_feature_signal`) averaged the zero placeholders of inactive tasks under active task balancing (`top_k=2` on a single active head with signal 6.0 reported `3.0`; `top_k=4` reported `1.5`), diluting sparse-label steps against dense ones and flipping which slot `worst_active` discards. #275/#276 fixed the identical code in the interaction learner only.

## New state

A shared `_active_topk_mean` implements the #276 semantics for both reducers: under task balancing, inactive rows are masked to `-inf`, `jax.lax.top_k` selects among active rows, and the mean divides by `min(#active, k)` with a guarded all-inactive branch; the `"none"` balancing branch stays byte-identical (`jnp.sort(...)[-k:]` mean).

Failing-test-first: `test_active_topk_excludes_inactive_zero_placeholders` fails on `main` and passes here; `test_active_topk_all_inactive_avoids_zero_division` (checkify float checks on both reducers) and `test_unbalanced_topk_still_averages_largest_heads` pin the guarded and unchanged branches.

## Evidence

Falsifier from #379 at this head: `top_k 1/2/4 -> [6.] [6.] [6.]`; ranking probe now `A -> 5.0, B -> 3.0` (matches the fixed interaction learner).

Verification at `d7bcea6979f00ff7308b41041d39201ce83fc7e8`:

```text
.venv/bin/python -m pytest tests/test_feature_discovery.py tests/test_interaction_task_utility_weights.py -q -o addopts=""   -> 110 passed
.venv/bin/python -m ruff check .                                                                                            -> All checks passed!
.venv/bin/python -m mypy                                                                                                    -> Success: no issues found in 164 source files
```

`core/feature_discovery.py` is not a registered evidence source (`core/interaction_features.py` is, and is untouched); no benchmark lane, seeds, or `outputs/` artifacts were touched; no `CHANGELOG.md` edit (maintainer-recorded).

CLI sessions cannot mint GitHub attachment URLs, so the run output above is inline rather than attached.

- [x] logs: inline above (focused pytest, ruff, mypy, falsifier output at the evidence head)

Provider/model/client: anthropic / claude-fable-5 / Claude Code 2.1.233 (contribute-to-asi skill revision 72e4239e).

### CODE REVIEW

**Review by lalalune — APPROVED** [review]
Independent review at exact head d7bcea6: READY. The fixed-budget learner now applies the already-reviewed active-top-k semantics to both reducers: inactive sentinels cannot dilute negative/positive active signals, all-inactive stays finite zero, active invalid values remain fail-visible, and the unbalanced branch remains unchanged. Verified in a clean current-main composition: 278 focused/adjacent tests passed, Ruff and strict mypy passed, and diff hygiene is clean. This source is not in the central five-claim registry; no output artifact changes.

**Review by lalalune — APPROVED** [review]
Approved at exact head d7bcea6. The shared reducer preserves the byte-equivalent unbalanced sort/top-k branch, masks inactive task rows before balanced selection, divides by min(active_count, k), and safely returns zero for all-inactive input. I adversarially checked signed active signals, inactive placeholders, inverse-frequency applicability, and immutable/JAX execution semantics. It composes with the current core queue: 377 affected tests passed, full Ruff passed, and strict mypy passed across 165 source files. No outputs or registered evidence sources changed.
